### ISSUE TITLE
Puppet-Lint integration on Windows does not work. 

### ISSUE THREAD

**Issue by myrlyn** [tropyx/NetBeansPuppet#24]

**Comment by myrlyn** [comment]
Execution fails with an IO Exception.
It seems to be looking for an executable with the name 'puppet-lint' rather than letting the automatic handling of ruby files occur.  
If I create a wrapper exe file named puppet-lint.exe that executes 
C:/Ruby23-x64/bin/ruby.exe C:/Ruby23-x64/lib/ruby/gems/2.3.0/gems/puppet-lint-2.0.0/bin/puppet-lint <ARGS>
and provides the output, and drop that somewhere in my %PATH% then Puppet-Lint seems to work correctly for the NetBeans plugin.

**Comment by mkleint** [comment]
> It seems to be looking for an executable with the name 'puppet-lint' rather than letting the automatic handling of ruby files occur.

yeah, I've never tested outside of my mac enviroment. What exactly does automatic handling of ruby files mean? Eg. what does the puppet lint related cmd line look like on windows? as the one above? how do I figure out the actual ruby and gem paths?

**Comment by myrlyn** [comment]
I'm not 100% sure where the bug is.  I suspect it has to do with the fact
that by default, Unix-like operating systems will attempt to run any file
with the execute permission as long as it is I  $PATH, while Windows looks
for a specific set of file extensions.
I'm not sure what the Ruby for Windows installer does to get gems to
execute correctly from the command-line, but when NetBeans tries to execute
the command it claims it cannot find the file.
On a hunch, I added an executable file named 'puppet-lint.exe' (which was a
simple program that calls the ruby executable on the pullet-lint script) to
a folder in %PATH%, and retried the NetBeans plug-in.   This resolved the
error.

**Comment by mkleint** [comment]
see https://github.com/tropyx/NetBeansPuppet/blob/master/src/main/java/com/tropyx/nb_puppet/lint/StatusProvider.java#L112 and https://github.com/tropyx/NetBeansPuppet/blob/master/src/main/java/com/tropyx/nb_puppet/lint/StatusProvider.java#L199

I wonder what the way to discover the executable on windows is. Does cygwin usage make things easier or more complicated?

**Comment by myrlyn** [comment]
as a cygwin user, I would guess probably 'more complicated'.
you'd probably have to launch NetBeans from a cygwin environment, which can
cause some confusion with Java, as it expects Windows file paths.
If you're going to try and bundle some kind of executable with the plugin,
I doubt Cygwin would be the way to go.

**Comment by mkleint** [comment]
I'll gladly accept patches but I don't have access to a windows machine myself unfortunately to test things out.

**Comment by myrlyn** [comment]
I'll take a look at it layer this week if I can.

**Comment by mkleint** [comment]
https://netbeans.org/bugzilla/show_bug.cgi?id=257907 is duplicate

**Comment by myrlyn** [comment]
I've not been able to get the plugin to build on my desktop - I get a unresolved dependencies from Maven, and I've not had time to chase it down. 
However, I have mocked up a findLint() method that *should* correctly detect the puppet-lint command on both windows and Mac, assuming ExternalProcessBuilder will handle bat files properly when presented the full path to the bat file.  This includes a stanza for checking /usr/local/bin on UnixLike machines (e.g. OSX, Linux) and in the current Ruby install dir on Windows. 

I'm not convinced this is the *correct* fix, as it re-implements the OS functionality of checking each PATH element in order for the executable. However, the actual bug seems to be in ExternalProcessBuilder, which is marked as Deprecated, so I doubt a bug request there will result in a fix.  

```java
public static String findLint() {
        //used later for special-casing PATH additions
        boolean unix = org.openide.util.Utilities.isMac() || org.openide.util.Utilities.isUnix();
        boolean win = org.openide.util.Utilities.isWindows();
        String paths[] = System.getenv("PATH").split(File.pathSeparator);
        //create an array with room for a PATH addition
        String pathext[] = new String[paths.length + 1];
        System.arraycopy(paths, 0, pathext, 1, paths.length);
        File temp;//File object for testing if we can execute the puppet-lint file in a given dir
        if (unix) {//add most common location to find puppet-lint that is not in $PATH
            pathext[0] = "/usr/local/bin/";
            for (String entry : pathext) {
                temp = new File(entry + File.separator + "puppet-lint");
                if (temp.canExecute()) {//check if the file exists and is executable
                    return temp.getAbsolutePath();//return the first match
                }
            }
        }
        if (win) {//add most common location to find puppet-lint that is not in $PATH
            pathext[0] = "C:\\Ruby23-x64\\bin";//current latest version of Ruby...
            for (String entry : pathext) {//two blocks here, because we check for two file names - .exe and .bat
                temp = new File(entry + File.separator + "puppet-lint.exe");
                if (temp.canExecute()) {
                    return temp.getAbsolutePath();
                }
                temp = new File(entry + File.separator + "puppet-lint.bat");
                if (temp.canExecute()) {
                    return temp.getAbsolutePath();
                }
            }
        }
        return "puppet-lint";//defer to the OS / old behavior
    }
```

**Comment by myrlyn** [comment]
Confirmed the above works on Windows (still unable to build the plugin, but replicated the issue in a mockup using ExternalProcessbuilder, then verified the above fixes the problem).
Another version that works, if you assume that puppet-lint will always be a .bat file on windows is the following, which is much simpler but may fail for some installations of puppet-lint: 
```java
   public static String findLint() {
        if (org.openide.util.Utilities.isMac()) {
            //mkleint: tired of tweaking the path of desktop apps to find
            // this location somehow on mac.
            if (new File("/usr/local/bin/puppet-lint").exists()) {
                return "/usr/local/bin/puppet-lint";
            }
        }
        if (org.openide.util.Utilities.isWindows()){
            //Deal with the fact that ExternalProcessBuilder does not check for .bat files by default
            return "puppet-lint.bat";
        }
        return "puppet-lint";
}
```

**Comment by myrlyn** [comment]
I submitted a pull request with a patch that should resolve this issue.

Jonathon Lee Walker

**Comment by mkleint** [comment]
thanks a lot, merged and published a new version (2.0.3)
